File: api/Endpoints/TrackArchiveCodec.cs
using System.IO.Compression;
using System.Text;

namespace HcWebApi.Endpoints
{
    /// <summary>
    /// One point of a runner's track as archived on their HC.HasherEventMap row.
    /// </summary>
    public sealed record ArchivedTrackPoint(
        long TimestampMs,
        double Latitude,
        double Longitude,
        double? Altitude,
        double? Accuracy,
        string? Type);

    /// <summary>
    /// Encodes a runner's track into the compact gzipped form stored in
    /// HC.HasherEventMap.TrackGzip (E5.F6.S4), and decodes it back.
    ///
    /// Format, version 1 — a gzip stream whose payload is:
    /// <code>
    ///   byte   version (1)
    ///   varint pointCount
    ///   per point, in ascending timestamp order:
    ///     zigzag-varint  Δ timestampMs        (first point: absolute)
    ///     zigzag-varint  Δ latitude  × 1e5    (first point: absolute)
    ///     zigzag-varint  Δ longitude × 1e5    (first point: absolute)
    ///     zigzag-varint  Δ altitude  × 10     (decimetres; 0 when unknown)
    ///     zigzag-varint  Δ accuracy  × 10     (decimetres; 0 when unknown)
    ///     varint         type byte length, then that many UTF-8 bytes (0 = no type)
    /// </code>
    /// Coordinates are already stored to 5 decimal places (≈1.1 m), so the
    /// integer scaling loses nothing. Consecutive points differ by a few metres
    /// and a few seconds, so every delta fits in one or two bytes before gzip
    /// squeezes the repetition further. A type string ("CHK", "OIN",
    /// "I-400.png::historic icehouse") travels verbatim so a replay can redraw
    /// every mark; a track without its marks cannot redraw the run.
    ///
    /// The version byte is the escape hatch: a reader that sees a version it
    /// does not know must refuse rather than guess.
    /// </summary>
    public static class TrackArchiveCodec
    {
        public const byte Version = 1;
        private const double CoordinateScale = 100_000.0;  // 1e-5 degrees
        private const double MetreScale = 10.0;            // decimetres

        public static byte[] Encode(IReadOnlyList<ArchivedTrackPoint> points)
        {
            using var raw = new MemoryStream();
            raw.WriteByte(Version);
            WriteVarint(raw, (ulong)points.Count);

            long prevTs = 0, prevLat = 0, prevLng = 0, prevAlt = 0, prevAcc = 0;
            foreach (ArchivedTrackPoint p in points)
            {
                long lat = (long)Math.Round(p.Latitude * CoordinateScale);
                long lng = (long)Math.Round(p.Longitude * CoordinateScale);
                long alt = p.Altitude.HasValue ? (long)Math.Round(p.Altitude.Value * MetreScale) : 0;
                long acc = p.Accuracy.HasValue ? (long)Math.Round(p.Accuracy.Value * MetreScale) : 0;

                WriteZigZag(raw, p.TimestampMs - prevTs);
                WriteZigZag(raw, lat - prevLat);
                WriteZigZag(raw, lng - prevLng);
                WriteZigZag(raw, alt - prevAlt);
                WriteZigZag(raw, acc - prevAcc);

                byte[] type = string.IsNullOrEmpty(p.Type) ? Array.Empty<byte>() : Encoding.UTF8.GetBytes(p.Type);
                WriteVarint(raw, (ulong)type.Length);
                raw.Write(type, 0, type.Length);

                prevTs = p.TimestampMs; prevLat = lat; prevLng = lng; prevAlt = alt; prevAcc = acc;
            }

            using var packed = new MemoryStream();
            using (var gzip = new GZipStream(packed, CompressionLevel.SmallestSize, leaveOpen: true))
            {
                raw.Position = 0;
                raw.CopyTo(gzip);
            }
            return packed.ToArray();
        }

        public static List<ArchivedTrackPoint> Decode(byte[] blob)
        {
            using var packed = new MemoryStream(blob);
            using var gzip = new GZipStream(packed, CompressionMode.Decompress);
            using var raw = new MemoryStream();
            gzip.CopyTo(raw);
            raw.Position = 0;

            int version = raw.ReadByte();
            if (version != Version)
            {
                throw new InvalidDataException($"TrackArchiveCodec: unknown version {version}.");
            }
            int count = checked((int)ReadVarint(raw));
            var points = new List<ArchivedTrackPoint>(count);

            long ts = 0, lat = 0, lng = 0, alt = 0, acc = 0;
            for (int i = 0; i < count; i++)
            {
                ts += ReadZigZag(raw);
                lat += ReadZigZag(raw);
                lng += ReadZigZag(raw);
                alt += ReadZigZag(raw);
                acc += ReadZigZag(raw);
                int typeLen = checked((int)ReadVarint(raw));
                string? type = null;
                if (typeLen > 0)
                {
                    byte[] typeBytes = new byte[typeLen];
                    int read = raw.Read(typeBytes, 0, typeLen);
                    if (read != typeLen) throw new InvalidDataException("TrackArchiveCodec: truncated type.");
                    type = Encoding.UTF8.GetString(typeBytes);
                }
                points.Add(new ArchivedTrackPoint(
                    ts,
                    lat / CoordinateScale,
                    lng / CoordinateScale,
                    alt == 0 ? null : alt / MetreScale,
                    acc == 0 ? null : acc / MetreScale,
                    type));
            }
            return points;
        }

        // --- varint plumbing (protobuf-style LEB128 with zigzag for signed values) ---

        private static void WriteVarint(Stream s, ulong value)
        {
            while (value >= 0x80)
            {
                s.WriteByte((byte)(value | 0x80));
                value >>= 7;
            }
            s.WriteByte((byte)value);
        }

        private static ulong ReadVarint(Stream s)
        {
            ulong result = 0;
            int shift = 0;
            while (true)
            {
                int b = s.ReadByte();
                if (b < 0) throw new InvalidDataException("TrackArchiveCodec: truncated varint.");
                result |= (ulong)(b & 0x7F) << shift;
                if ((b & 0x80) == 0) return result;
                shift += 7;
                if (shift > 63) throw new InvalidDataException("TrackArchiveCodec: varint too long.");
            }
        }

        private static void WriteZigZag(Stream s, long value) =>
            WriteVarint(s, (ulong)((value << 1) ^ (value >> 63)));

        private static long ReadZigZag(Stream s)
        {
            ulong u = ReadVarint(s);
            return (long)(u >> 1) ^ -(long)(u & 1);
        }
    }
}

File: mobile-app/lib/util/track_archive_codec.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

/// One point of an archived PackTrack trail, as the server stored it.
class ArchivedTrackPoint {
  const ArchivedTrackPoint({
    required this.timestampMs,
    required this.lat,
    required this.lng,
    this.alt,
    this.acc,
    this.type,
  });

  final int timestampMs;
  final double lat;
  final double lng;
  final double? alt;
  final double? acc;
  final String? type;
}

/// Thrown when a blob is not a version this app can read. The caller falls
/// back to the server for that run rather than drawing anything.
class TrackArchiveVersionException implements Exception {
  const TrackArchiveVersionException(this.version);
  final int version;
  @override
  String toString() => 'TrackArchiveCodec: unknown version $version';
}

/// Decoder for `HasherEventMap.TrackGzip` — a faithful port of the API's
/// `TrackArchiveCodec.Decode` (E5.F6.S4 / E5.F7.S1). The blob is a gzip
/// stream whose payload is:
///
/// ```
///   byte   version (1)
///   varint pointCount
///   per point, ascending by time:
///     zigzag-varint  Δ timestampMs        (first point: absolute)
///     zigzag-varint  Δ latitude  × 1e5
///     zigzag-varint  Δ longitude × 1e5
///     zigzag-varint  Δ altitude  × 10     (decimetres; 0 = unknown)
///     zigzag-varint  Δ accuracy  × 10     (decimetres; 0 = unknown)
///     varint         type byte length, then that many UTF-8 bytes
/// ```
///
/// Verified against production blobs in `test/track_archive_codec_test.dart`
/// (the C# decoder's output is the expected value). Decode only: the phone
/// never writes an archive.
class TrackArchiveCodec {
  TrackArchiveCodec._();

  static const int version = 1;
  static const double _coordinateScale = 1e5;
  static const double _metreScale = 10;

  /// The sync carries the VARBINARY as base64 text; this is what the local
  /// `trackGzip` column holds.
  static List<ArchivedTrackPoint> decodeBase64(String base64Blob) =>
      decode(base64Decode(base64Blob.trim()));

  static List<ArchivedTrackPoint> decode(List<int> blob) {
    final Uint8List raw = Uint8List.fromList(gzip.decode(blob));
    final _Reader r = _Reader(raw);
    final int v = r.byte();
    if (v != version) throw TrackArchiveVersionException(v);
    final int count = r.varint();
    final List<ArchivedTrackPoint> points = <ArchivedTrackPoint>[];
    int ts = 0, lat = 0, lng = 0, alt = 0, acc = 0;
    for (int i = 0; i < count; i++) {
      ts += r.zigzag();
      lat += r.zigzag();
      lng += r.zigzag();
      alt += r.zigzag();
      acc += r.zigzag();
      final int typeLen = r.varint();
      String? type;
      if (typeLen > 0) {
        type = utf8.decode(r.bytes(typeLen));
      }
      points.add(
        ArchivedTrackPoint(
          timestampMs: ts,
          lat: lat / _coordinateScale,
          lng: lng / _coordinateScale,
          alt: alt == 0 ? null : alt / _metreScale,
          acc: acc == 0 ? null : acc / _metreScale,
          type: type,
        ),
      );
    }
    return points;
  }
}

class _Reader {
  _Reader(this._b);
  final Uint8List _b;
  int _p = 0;

  int byte() {
    if (_p >= _b.length) throw const FormatException('TrackArchiveCodec: truncated');
    return _b[_p++];
  }

  /// Unsigned LEB128, as the C# writer emits it (up to 10 bytes for 64 bits).
  int varint() {
    int result = 0;
    int shift = 0;
    while (true) {
      final int b = byte();
      result |= (b & 0x7f) << shift;
      if (b & 0x80 == 0) return result;
      shift += 7;
      if (shift > 63) throw const FormatException('TrackArchiveCodec: varint too long');
    }
  }

  /// Zigzag: (n << 1) ^ (n >> 63) on the way in, so (u >>> 1) ^ -(u & 1) out.
  int zigzag() {
    final int u = varint();
    return (u >>> 1) ^ -(u & 1);
  }

  Uint8List bytes(int n) {
    if (_p + n > _b.length) throw const FormatException('TrackArchiveCodec: truncated type');
    final Uint8List out = Uint8List.sublistView(_b, _p, _p + n);
    _p += n;
    return out;
  }
}
